packages/arb-avm-cpp/cavm/ccheckpointstorage.cpp

```diff
@@ -130,68 +127,75 @@ int deleteValue(CCheckpointStorage* storage_ptr, void* hash_key) {
 
     auto results = deleter.deleteValue(hash_key_vector);
 
-    if (results.status.ok()) {
-        auto status = deleter.commitTransaction();
-        return status.ok();
-    } else {
-        false;
+    if (!results.status.ok()) {
+        return false;
     }
+
+    auto status = deleter.commitTransaction();
+    return status.ok();
 }
 
 int saveData(CCheckpointStorage* storage_ptr,
-             const char* key,
-             const char* value) {
+             const void* key,
+             const void* data) {
     auto storage = static_cast<CheckpointStorage*>(storage_ptr);
     auto transaction = storage->makeTransaction();
 
-    auto key_str = std::string(key);
+    auto key_ptr = reinterpret_cast<const char*>(key);
+    auto data_ptr = reinterpret_cast<const char*>(data);
+
+    auto key_str = std::string(key_ptr);
+    auto data_str = std::string(data_ptr);
+
     auto key_vector =
         std::vector<unsigned char>(key_str.begin(), key_str.end());
+    auto data_vector =
+        std::vector<unsigned char>(data_str.begin(), data_str.end());
```

Comment:
This is super unsafe. The `std::string` constructor taking a `const char*` assumes that the pointer is to a c string which is null terminated. The go calls this with the pointer to the beginning of a byte slice which is not null terminated. Instead you should also pass the key and value lengths and then construct the vectors directly from the pointers.
ex: `auto key_vector = std::vector<unsigned char>(key, key + key_length);`